File: server-sdks/sockudo-http-go/push.go
package sockudo

import (
	"encoding/json"
	"fmt"
)

type PushCursorParams struct {
	Limit  *int
	Cursor *string
}

func (params PushCursorParams) toMap() map[string]string {
	m := make(map[string]string)
	if params.Limit != nil {
		m["limit"] = fmt.Sprintf("%d", *params.Limit)
	}
	if params.Cursor != nil {
		m["cursor"] = *params.Cursor
	}
	return m
}

type PushSubscriptionParams struct {
	Limit    *int
	Cursor   *string
	Channel  *string
	DeviceID *string
}

func (params PushSubscriptionParams) toMap() map[string]string {
	m := PushCursorParams{Limit: params.Limit, Cursor: params.Cursor}.toMap()
	if params.Channel != nil {
		m["channel"] = *params.Channel
	}
	if params.DeviceID != nil {
		m["deviceId"] = *params.DeviceID
	}
	return m
}

type PushRecipient map[string]interface{}
type PushDeviceDetails map[string]interface{}
type PushChannelSubscription map[string]interface{}
type PushPublishRequest map[string]interface{}
type PushDeliveryStatusEvent map[string]interface{}

func pushHeaders(capability string, deviceIdentityToken string) map[string]string {
	if capability == "" {
		capability = "push-admin"
	}
	headers := map[string]string{"X-Sockudo-Push-Capability": capability}
	if deviceIdentityToken != "" {
		headers["X-Sockudo-Device-Identity-Token"] = deviceIdentityToken
	}
	return headers
}

func (c *Client) pushRequest(method, path string, body interface{}, params map[string]string, extraHeaders map[string]string) ([]byte, error) {
	var payload []byte
	var err error
	if body != nil {
		payload, err = json.Marshal(body)
		if err != nil {
			return nil, err
		}
	}

	requestURL, err := createRequestURL(method, c.Host, fmt.Sprintf("/apps/%s/push%s", c.AppID, path), c.Key, c.Secret, authTimestamp(), c.Secure, payload, params, c.Cluster)
	if err != nil {
		return nil, err
	}
	return c.requestWithExtraHeaders(method, requestURL, payload, extraHeaders)
}

func (c *Client) ActivateDevice(device PushDeviceDetails) ([]byte, error) {
	return c.pushRequest("POST", "/deviceRegistrations", device, nil, pushHeaders("push-admin", ""))
}

func (c *Client) CreateDeviceActivation(device PushDeviceDetails) ([]byte, error) {
	return c.ActivateDevice(device)
}

func (c *Client) UpdateDeviceRegistration(device PushDeviceDetails, deviceIdentityToken string) ([]byte, error) {
	return c.pushRequest("POST", "/deviceRegistrations", device, nil, pushHeaders("push-subscribe", deviceIdentityToken))
}

func (c *Client) ListDeviceRegistrations(params PushCursorParams) ([]byte, error) {
	return c.pushRequest("GET", "/deviceRegistrations", nil, params.toMap(), pushHeaders("push-admin", ""))
}

func (c *Client) GetDeviceRegistration(deviceID string, deviceIdentityToken string) ([]byte, error) {
	capability := "push-admin"
	if deviceIdentityToken != "" {
		capability = "push-subscribe"
	}
	return c.pushRequest("GET", "/deviceRegistrations/"+deviceID, nil, nil, pushHeaders(capability, deviceIdentityToken))
}

func (c *Client) DeleteDeviceRegistration(deviceID string, deviceIdentityToken string) ([]byte, error) {
	capability := "push-admin"
	if deviceIdentityToken != "" {
		capability = "push-subscribe"
	}
	return c.pushRequest("DELETE", "/deviceRegistrations/"+deviceID, nil, nil, pushHeaders(capability, deviceIdentityToken))
}

func (c *Client) RemoveDeviceRegistrationsByClient(clientID string) ([]byte, error) {
	return c.pushRequest("DELETE", "/deviceRegistrations", nil, map[string]string{"clientId": clientID}, pushHeaders("push-admin", ""))
}

func (c *Client) UpsertChannelPushSubscription(subscription PushChannelSubscription, deviceIdentityToken string) ([]byte, error) {
	capability := "push-admin"
	if deviceIdentityToken != "" {
		capability = "push-subscribe"
	}
	return c.pushRequest("POST", "/channelSubscriptions", subscription, nil, pushHeaders(capability, deviceIdentityToken))
}

func (c *Client) ListChannelPushSubscriptions(params PushSubscriptionParams, deviceIdentityToken string) ([]byte, error) {
	capability := "push-admin"
	if deviceIdentityToken != "" {
		capability = "push-subscribe"
	}
	return c.pushRequest("GET", "/channelSubscriptions", nil, params.toMap(), pushHeaders(capability, deviceIdentityToken))
}

func (c *Client) DeleteChannelPushSubscriptions(params PushSubscriptionParams, deviceIdentityToken string) ([]byte, error) {
	capability := "push-admin"
	if deviceIdentityToken != "" {
		capability = "push-subscribe"
	}
	return c.pushRequest("DELETE", "/channelSubscriptions", nil, params.toMap(), pushHeaders(capability, deviceIdentityToken))
}

func (c *Client) ListChannelPushSubscriptionChannels(params PushCursorParams) ([]byte, error) {
	return c.pushRequest("GET", "/channelSubscriptions/channels", nil, params.toMap(), pushHeaders("push-admin", ""))
}

func (c *Client) ListPushCredentials(params PushCursorParams) ([]byte, error) {
	return c.pushRequest("GET", "/credentials", nil, params.toMap(), pushHeaders("push-admin", ""))
}

func (c *Client) PutPushCredential(provider string, credential map[string]interface{}) ([]byte, error) {
	return c.pushRequest("POST", "/credentials/"+provider, credential, nil, pushHeaders("push-admin", ""))
}

func (c *Client) PublishPush(request PushPublishRequest) ([]byte, error) {
	request["sync"] = false
	return c.pushRequest("POST", "/publish", request, nil, pushHeaders("push-admin", ""))
}

func (c *Client) PublishPushDirect(request PushPublishRequest) ([]byte, error) {
	return c.PublishPush(request)
}

func (c *Client) PublishPushBatch(requests []PushPublishRequest) ([]byte, error) {
	for i := range requests {
		requests[i]["sync"] = false
	}
	return c.pushRequest("POST", "/batch/publish", requests, nil, pushHeaders("push-admin", ""))
}

func (c *Client) SchedulePush(request PushPublishRequest) ([]byte, error) {
	if _, ok := request["notBeforeMs"]; !ok {
		return nil, fmt.Errorf("scheduled push requires notBeforeMs")
	}
	return c.PublishPush(request)
}

func (c *Client) GetPublishStatus(publishID string) ([]byte, error) {
	return c.pushRequest("GET", "/publish/"+publishID+"/status", nil, nil, pushHeaders("push-admin", ""))
}

func (c *Client) CancelScheduledPush(publishID string) ([]byte, error) {
	return c.pushRequest("DELETE", "/scheduled/"+publishID, nil, nil, pushHeaders("push-admin", ""))
}

func (c *Client) PostPushDeliveryStatus(event PushDeliveryStatusEvent) ([]byte, error) {
	return c.pushRequest("POST", "/deliveryStatus", event, nil, pushHeaders("push-admin", ""))
}

File: server-sdks/sockudo-http-rust/src/push.rs
use crate::{Result, Sockudo, SockudoError};
use reqwest::Response;
use serde::{Deserialize, Serialize};
use sonic_rs::{JsonValueTrait, Value, json};
use std::collections::{BTreeMap, HashMap};

#[derive(Clone, Debug, Default, Serialize, Deserialize)]
pub struct PushCursorParams {
    pub limit: Option<u32>,
    pub cursor: Option<String>,
}

impl PushCursorParams {
    pub fn to_map(&self) -> BTreeMap<String, String> {
        let mut map = BTreeMap::new();
        if let Some(limit) = self.limit {
            map.insert("limit".to_string(), limit.to_string());
        }
        if let Some(cursor) = &self.cursor {
            map.insert("cursor".to_string(), cursor.clone());
        }
        map
    }
}

#[derive(Clone, Debug, Default, Serialize, Deserialize)]
pub struct PushSubscriptionParams {
    pub limit: Option<u32>,
    pub cursor: Option<String>,
    pub channel: Option<String>,
    pub device_id: Option<String>,
}

impl PushSubscriptionParams {
    pub fn to_map(&self) -> BTreeMap<String, String> {
        let mut map = PushCursorParams {
            limit: self.limit,
            cursor: self.cursor.clone(),
        }
        .to_map();
        if let Some(channel) = &self.channel {
            map.insert("channel".to_string(), channel.clone());
        }
        if let Some(device_id) = &self.device_id {
            map.insert("deviceId".to_string(), device_id.clone());
        }
        map
    }
}

fn push_headers(capability: &str, device_identity_token: Option<&str>) -> HashMap<String, String> {
    let mut headers = HashMap::from([(
        "X-Sockudo-Push-Capability".to_string(),
        capability.to_string(),
    )]);
    if let Some(token) = device_identity_token {
        headers.insert(
            "X-Sockudo-Device-Identity-Token".to_string(),
            token.to_string(),
        );
    }
    headers
}

fn push_path(path: &str) -> String {
    format!("/push{path}")
}

impl Sockudo {
    pub async fn activate_device(&self, device: &Value) -> Result<Response> {
        self.post_with_headers(
            &push_path("/deviceRegistrations"),
            device,
            &push_headers("push-admin", None),
        )
        .await
    }

    pub async fn create_device_activation(&self, device: &Value) -> Result<Response> {
        self.activate_device(device).await
    }

    pub async fn update_device_registration(
        &self,
        device: &Value,
        device_identity_token: &str,
    ) -> Result<Response> {
        self.post_with_headers(
            &push_path("/deviceRegistrations"),
            device,
            &push_headers("push-subscribe", Some(device_identity_token)),
        )
        .await
    }

    pub async fn list_device_registrations(
        &self,
        params: Option<&PushCursorParams>,
    ) -> Result<Response> {
        let params = params.map(PushCursorParams::to_map);
        self.get_with_headers(
            &push_path("/deviceRegistrations"),
            params.as_ref(),
            &push_headers("push-admin", None),
        )
        .await
    }

    pub async fn get_device_registration(
        &self,
        device_id: &str,
        device_identity_token: Option<&str>,
    ) -> Result<Response> {
        let capability = if device_identity_token.is_some() {
            "push-subscribe"
        } else {
            "push-admin"
        };
        self.get_with_headers(
            &push_path(&format!("/deviceRegistrations/{device_id}")),
            None,
            &push_headers(capability, device_identity_token),
        )
        .await
    }

    pub async fn delete_device_registration(
        &self,
        device_id: &str,
        device_identity_token: Option<&str>,
    ) -> Result<Response> {
        let capability = if device_identity_token.is_some() {
            "push-subscribe"
        } else {
            "push-admin"
        };
        self.delete_with_headers(
            &push_path(&format!("/deviceRegistrations/{device_id}")),
            None,
            &push_headers(capability, device_identity_token),
        )
        .await
    }

    pub async fn remove_device_registrations_by_client(&self, client_id: &str) -> Result<Response> {
        let params = BTreeMap::from([("clientId".to_string(), client_id.to_string())]);
        self.delete_with_headers(
            &push_path("/deviceRegistrations"),
            Some(&params),
            &push_headers("push-admin", None),
        )
        .await
    }

    pub async fn upsert_channel_push_subscription(
        &self,
        subscription: &Value,
        device_identity_token: Option<&str>,
    ) -> Result<Response> {
        let capability = if device_identity_token.is_some() {
            "push-subscribe"
        } else {
            "push-admin"
        };
        self.post_with_headers(
            &push_path("/channelSubscriptions"),
            subscription,
            &push_headers(capability, device_identity_token),
        )
        .await
    }

    pub async fn list_channel_push_subscriptions(
        &self,
        params: Option<&PushSubscriptionParams>,
        device_identity_token: Option<&str>,
    ) -> Result<Response> {
        let capability = if device_identity_token.is_some() {
            "push-subscribe"
        } else {
            "push-admin"
        };
        let params = params.map(PushSubscriptionParams::to_map);
        self.get_with_headers(
            &push_path("/channelSubscriptions"),
            params.as_ref(),
            &push_headers(capability, device_identity_token),
        )
        .await
    }

    pub async fn delete_channel_push_subscriptions(
        &self,
        params: &PushSubscriptionParams,
        device_identity_token: Option<&str>,
    ) -> Result<Response> {
        let capability = if device_identity_token.is_some() {
            "push-subscribe"
        } else {
            "push-admin"
        };
        let params = params.to_map();
        self.delete_with_headers(
            &push_path("/channelSubscriptions"),
            Some(&params),
            &push_headers(capability, device_identity_token),
        )
        .await
    }

    pub async fn list_channel_push_subscription_channels(
        &self,
        params: Option<&PushCursorParams>,
    ) -> Result<Response> {
        let params = params.map(PushCursorParams::to_map);
        self.get_with_headers(
            &push_path("/channelSubscriptions/channels"),
            params.as_ref(),
            &push_headers("push-admin", None),
        )
        .await
    }

    pub async fn list_push_credentials(
        &self,
        params: Option<&PushCursorParams>,
    ) -> Result<Response> {
        let params = params.map(PushCursorParams::to_map);
        self.get_with_headers(
            &push_path("/credentials"),
            params.as_ref(),
            &push_headers("push-admin", None),
        )
        .await
    }

    pub async fn put_push_credential(
        &self,
        provider: &str,
        credential: &Value,
    ) -> Result<Response> {
        self.post_with_headers(
            &push_path(&format!("/credentials/{provider}")),
            credential,
            &push_headers("push-admin", None),
        )
        .await
    }

    pub async fn publish_push(&self, request: &Value) -> Result<Response> {
        let mut request = request.clone();
        request["sync"] = json!(false);
        self.post_with_headers(
            &push_path("/publish"),
            &request,
            &push_headers("push-admin", None),
        )
        .await
    }

    pub async fn publish_push_direct(&self, request: &Value) -> Result<Response> {
        self.publish_push(request).await
    }

    pub async fn publish_push_batch(&self, requests: &[Value]) -> Result<Response> {
        let requests: Vec<Value> = requests
            .iter()
            .map(|request| {
                let mut item = request.clone();
                item["sync"] = json!(false);
                item
            })
            .collect();
        let body = sonic_rs::to_value(&requests).map_err(SockudoError::Json)?;
        self.post_with_headers(
            &push_path("/batch/publish"),
            &body,
            &push_headers("push-admin", None),
        )
        .await
    }

    pub async fn schedule_push(&self, request: &Value) -> Result<Response> {
        if request.get("notBeforeMs").is_none() {
            return Err(SockudoError::Validation {
                message: "scheduled push requires notBeforeMs".to_string(),
            });
        }
        self.publish_push(request).await
    }

    pub async fn get_publish_status(&self, publish_id: &str) -> Result<Response> {
        self.get_with_headers(
            &push_path(&format!("/publish/{publish_id}/status")),
            None,
            &push_headers("push-admin", None),
        )
        .await
    }

    pub async fn cancel_scheduled_push(&self, publish_id: &str) -> Result<Response> {
        self.delete_with_headers(
            &push_path(&format!("/scheduled/{publish_id}")),
            None,
            &push_headers("push-admin", None),
        )
        .await
    }

    pub async fn post_push_delivery_status(&self, event: &Value) -> Result<Response> {
        self.post_with_headers(
            &push_path("/deliveryStatus"),
            event,
            &push_headers("push-admin", None),
        )
        .await
    }
}
